src/themis/secure_cell_seal_passphrase.c

```diff
@@ -0,0 +1,341 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "themis/secure_cell_seal_passphrase.h"
+
+#include "soter/soter_kdf.h"
+#include "soter/soter_rand.h"
+#include "soter/soter_wipe.h"
+
+#include "themis/secure_cell_alg.h"
+#include "themis/sym_enc_message.h"
+
+/*
+ * Themis always uses the default algorithm and parameters for encryption.
+ * These may be transparently updated between library releases. However,
+ * decryption code has to support all previously produces data formats.
+ */
+
+/* Must be a #define to be treated as compile-time constant */
+#define DEFAULT_AUTH_TOKEN_SIZE                                                                                \
+    (themis_scell_auth_token_passphrase_min_size + THEMIS_AUTH_SYM_IV_LENGTH + THEMIS_AUTH_SYM_AUTH_TAG_LENGTH \
+     + themis_scell_pbkdf2_context_min_size + THEMIS_AUTH_SYM_PBKDF2_SALT_LENGTH)
+
+static inline uint32_t soter_alg_without_kdf(uint32_t alg)
+{
+    /* Explicit cast for mask to have correct size before negation */
+    return (alg & ~((uint32_t)SOTER_SYM_KDF_MASK)) | SOTER_SYM_NOKDF;
+}
+
+static inline size_t soter_alg_key_length(uint32_t alg)
+{
+    return (alg & SOTER_SYM_KEY_LENGTH_MASK) / 8;
+}
+
+static inline bool soter_alg_reserved_bits_valid(uint32_t alg)
+{
+    static const uint32_t used_bits = SOTER_SYM_KEY_LENGTH_MASK | SOTER_SYM_PADDING_MASK
+                                      | SOTER_SYM_KDF_MASK | SOTER_SYM_ALG_MASK;
+    return (alg & ~used_bits) == 0;
+}
+
+themis_status_t themis_auth_sym_encrypt_message_with_passphrase_(const uint8_t* passphrase,
+                                                                 size_t passphrase_length,
+                                                                 const uint8_t* message,
+                                                                 size_t message_length,
+                                                                 const uint8_t* user_context,
+                                                                 size_t user_context_length,
+                                                                 uint8_t* auth_token,
+                                                                 size_t* auth_token_length,
+                                                                 uint8_t* encrypted_message,
+                                                                 size_t* encrypted_message_length)
+{
+    themis_status_t res = THEMIS_FAIL;
```

Comment:
> for what declaration all variables at start of function?

That's mostly my (faint) hopes of having native Windows support one day. **cl.exe** _still_ supports only C89 with some Microsoft extensions, and fails to compile code that declares variables between expressions. Moving all those variables will be a hassle if we decide to support MSVC so I figured that we might just write the code C89-style from the get-go.

I strive to keep functions short enough – one screen of code – so that it's possible to see all logic *and* variables in one glance. Admittedly,  I‘m not always successful at that :(

i've found that this style does not hurt readability much in most cases, especially if the functions are kept short.

Reply:
it's make sense for windows support. thanks, I didn't know about that